### DISCUSSION TITLE
Help me to interpret the error message from lax.cond

### DISCUSSION THREAD

**Discussion by ziyuanzhao2000** [jax-ml/jax#13411] [Q&A] [answered]
Below is a minimal code snippet from a reinforcement learning code I wrote:

```
from typing import *
from dataclasses import dataclass
import jax 
import jax.numpy as jnp
from jax.random import PRNGKey
from jax._src.typing import ArrayLike
from functools import partial
from jax import tree_util, lax

class State(NamedTuple):
    x: int
    y: int

    def _tree_flatten(self):
        return ((), {'x': self.x, 'y': self.y})
    
    @classmethod
    def _tree_unflatten(cls, aux_data, children):
        return cls(*children, **aux_data)

for c in [State]:
    tree_util.register_pytree_node(c,
                                   c._tree_flatten,
                                   c._tree_unflatten)

A = State(1, 1)
B = State(1, 2)


# augment = lambda f: jax.jit((lambda s: lax.cond(A==s,
#                                                 lambda _: B,
#                                                 lambda s: f(s), 
#                                                 s)))

augment = lambda f: (lambda s: lax.cond(A==s,
                                        lambda _: B,
                                        lambda s: f(s), 
                                        s))

left = augment(lambda s: State(max(0, s.x-1), s.y))

left(B)

```

However, it's giving me the error of:
```
jax._src.traceback_util.UnfilteredStackTrace: TypeError: true_fun and false_fun output must have same type structure, got PyTreeDef(CustomNode(State[{'x': 1, 'y': 2}], [])) and PyTreeDef(CustomNode(State[{'x': 0, 'y': 2}], [])).
```

But I can't see why they do not have the same structure...

**Comment by jakevdp** [comment, marked as answer]
The reason you get this error is because your `tree_flatten` function treats the `x` and `y` values as static, and so the actual values of `x` and `y` must match in both branches.

If you want the `x` and `y` values to be dynamic, you can include them in `children` rather than `aux_data`:
```python
class State(NamedTuple):
    x: int
    y: int

    def _tree_flatten(self):
        return ((self.x, self.y), {})
    
    @classmethod
    def _tree_unflatten(cls, aux_data, children):
        return cls(*children, **aux_data)
```
and then since the values are traced, you can use `jnp.maximum` in place of the built-in `max` function:
```python
left = augment(lambda s: State(jnp.maximum(0, s.x-1), s.y))
```
If this solution suits you, note that you could remove the `tree_flatten`/`tree_unflatten`/`register_pytree_node` steps for `State` altogether, because `NamedTuple` is registered by default with this precise behavior (i.e. array-like attributes are considered dynamic).

**Comment by ziyuanzhao2000** [comment]
Hi Jake, thanks for your answer! Unfortunately my code still doesn't work for some other reason...and now I see error like
```
jax._src.errors.ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(int32[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
The axis argument must be known statically.
The error occurred while tracing the function <lambda> at /.../jax_debug.py:37 for cond. This concrete value was not available in Python because it depends on the value of the argument 's'.

See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.ConcretizationTypeError
```

which I'm not sure how to fix. (I thought using lax.cond already solves the problem)

**Reply by jakevdp** [reply]
It looks like you're passing a traced value to the `axis` argument of some function. Generally axes must be static. There's not much more I can say without seeing the code.

Edit: is it possible you used `jnp.max` instead of `jnp.maximum`? That would lead to this error, because the second argument to `jnp.max` is the axis.

**Reply by ziyuanzhao2000** [reply]
```
augment = lambda f: (lambda s: B if A==s else f(s))

left = augment(jax.jit(lambda s: State(max(0, s.x-1), s.y)))

left(B)
```

This was my code when it gives that error. So I haven't used `jnp.max` yet. To debug I even take away lax.cond and just try to jit compile the internal (pure) function that transform my state. Still it seems that jax is unhappy for some reason.

**Reply by jakevdp** [reply]
Did you see my note about `jnp.maximum` in the original answer?

**Reply by ziyuanzhao2000** [reply]
Yes, I tried `jnp.max` before discarding it, oh wait, so I just realized that you are referring to `jnp.maximum` rather than `jnp.max,` and that fixes the problem. I used `jnp.max` before and that messed up things.

**Reply by ziyuanzhao2000** [reply]
That's interesting! So in general I should use `jax.numpy` version of functions whenever possible?

By the way, it seems my `lax.cond` code that substitutes the conditional won't work. Maybe I also misunderstood how to use `lax.cond`? Thanks for answering my basic questions anyway! :)
